src/lib/download-progress.ts

```diff
@@ -0,0 +1,88 @@
+import * as _ from 'lodash';
+import blink = require('./blink');
+import { DeviceStatus, TargetState } from '../types/state';
+import * as deviceState from '../device-state';
+import * as targetState from '../device-state/target-state';
+
+interface ServiceState {
+	id: string;
+	status: string;
+	download_progress: number | null;
+}
+
+const DOWNLOAD_BLINK_THROTTLE = 10000;
+const DOWNLOAD_BLINK_PERIOD_BASE = 100;
+const DOWNLOAD_BLINK_PERIOD_STEP = 10;
+
+const throttledBlink = _.throttle((progress: number) => {
+	blink.pattern.stop();
+	if (progress < 100) {
+		blink.pattern.start({
+			blinks: 1,
+			pause: 0,
+			onDuration:
+				DOWNLOAD_BLINK_PERIOD_BASE + DOWNLOAD_BLINK_PERIOD_STEP * progress,
+			offDuration:
+				DOWNLOAD_BLINK_PERIOD_BASE + DOWNLOAD_BLINK_PERIOD_STEP * progress,
+		});
+	}
+}, DOWNLOAD_BLINK_THROTTLE);
+
+export async function downloadProgress() {
+	const currentDeviceState = await deviceState.getStatus();
+
+	if (currentDeviceState.local?.update_pending) {
+		// Get target services and current state
+		const targetDeviceState = await targetState.get();
```

Comment:
I think you might be able to use `deviceState.getTarget` and skip an import

Reply:
Ah good call, i was not aware of that method.